File: librarian/services/book_service.py
class book_service:
    def __init__(self, book_repo, borrow_data_repo):
        self.book_repo = book_repo
        self.borrow_data_repo = borrow_data_repo


    def validate(self, data):
        if not data:
            raise Exception("data is null")

        if not data.isbn:
            raise Exception("isbn kosong")

        if not data.nama:
            raise Exception("nama kosong")

        if not data.pengarang:
            raise Exception("pengarang kosong")

        if not data.tahun_terbit:
            raise Exception("tahun terbit kosong")

        if data.qty < 0:
            raise Exception("jumlah buku invalid")

        dupe = self.book_repo.get(data.isbn)

        if dupe:
            raise Exception("isbn sudah terdaftar")

    
    def create(self, data):
        # Validating.
        self.validate(data)

        # Adding.
        self.book_repo.create(data)


    def update(self, data):
        # Validating.
        self.validate(data)

        # Updating.
        self.book_repo.update(data)


    def delete(self, isbn):
        # Validating.
        if not isbn:
            raise Exception("isbn kosong")

        # Deleting from borrow repo.
        self.borrow_data_repo.delete_by_isbn(isbn)

        # Deleting from book repo.
        self.book_repo.delete(isbn)

File: internet-programming-uts1/src/logic/services/BookService.java
package logic.services;

import logic.models.Book;
import logic.models.BookCategory;
import logic.repositories.IBookCategoryRepository;
import logic.repositories.IBookRepository;
import logic.repositories.IOrderRepository;

public class BookService 
{
    IBookRepository _bookRepo;
    IBookCategoryRepository _catRepo;
    IOrderRepository _orderRepo;

    public BookService(
        IBookRepository bookRepo,
        IBookCategoryRepository catRepo,
        IOrderRepository orderRepo)
    {
        _bookRepo = bookRepo;
        _catRepo = catRepo;
        _orderRepo = orderRepo;
    }

    void validate(Book data)
    throws Exception
    {
        if (data == null)
        {
            throw new Exception("data is null");
        }

        BookCategory cat = _catRepo.get(data.catId);

        if (cat == null)
        {
            throw new Exception("invalid category");
        }

        if (data.title == null ||
            data.title.trim().isEmpty())
        {
            throw new Exception("title is empty");
        }

        if (data.writer == null ||
            data.writer.trim().isEmpty())
        {
            throw new Exception("writer is empty");
        }

        if (data.price < 0)
        {
            throw new Exception("invalid price");
        }
    }

    public Book create(Book data)
    throws Exception
    {
        // Validating.
        validate(data);

        // Adding.
        return _bookRepo.create(data);
    }

    public void update(Book data)
    throws Exception
    {
        // Validating.
        validate(data);

        // Updating.
        _bookRepo.update(data);
    }

    public void delete(int id)
    throws Exception
    {
        // Deleting orders.
        _orderRepo.deleteByBook(id);

        // Deleting book.
        _bookRepo.delete(id);
    }
}
